### DISCUSSION TITLE
Usage of `inline` attribute

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2797] [General]
We discussed this in the past when Kix was about but I can't find it. From memory his suggestion was to mentally convert functions to assembler, take into consideration branching, and then add `#[inline]` to "small" functions. Even if that was a well defined algorithm, and assuming I could do that, it is way too much work to do repo wide.

Can we come up with some more realistic hueristic for when to apply the `#[inline]` attribute so that we can start using it more uniformly?

**Comment by tcharding** [comment]
Related SO post: https://stackoverflow.com/questions/37639276/when-should-inline-be-used-in-rust

**Comment by tcharding** [comment]
If we can, @jamillambert is keen to get his patching skills up and may be willing to do it.

**Comment by dpc** [comment]
IMO, just don't bother, unless you have a benchmark showing that it is making significant difference. There are probably bigger fishes to fry, and optimizing can always be done later, after people actually identify optimization opportunities worth pursuing.

**Reply by storopoli** [reply]
I agree, we should maybe expand the `cargo bench` suite (I haven't had the opportunity to check the benchmarks part of the codebase yet) and `#[inline]` only if necessary and we see evidence in benchmarks that inlining would be beneficial.

**Reply by apoelstra** [reply]
I agree, I think "if you can measure it in a benchmark" is the only sane rule that we can reliably define and use.

And yes, if @jamillambert is looking to practice writing patches, adding benchmarks would be awesome :)

**Reply by TheBlueMatt** [reply]
You can't benchmark it directly, though, because `#[inline]` is more for downstream crates than for use inside the same crate (https://github.com/rust-bitcoin/rust-bitcoin/discussions/2797#discussioncomment-9917932)

**Comment by apoelstra** [comment]
I wouldn't mind removing #[inline] from existing functions in the crate, unless they are literally one or two (logical) lines of code.

In general I suspect we overuse inline here. We rarely get complaints about runtime performance and when we do it's because of big things like not using vectorization or CPU extensions in our sha2 code. Meanwhile, at least in rust-miniscript, we often get complaints about compiler time and binary size.

**Reply by Kixunil** [reply]
I personally think most of the places are correct usage. Compile times will improve with splitting the library much better than with `#[inline]` fiddling, I think.

**Reply by apoelstra** [reply]
Definitely agreed that splitting the library will be more helpful than `#[inline]` fiddling.

I also agree that "most of" the uses are correct :).

**Comment by Kixunil** [comment]
One case I think should always have `#[inline]` is functions with panicking preconditions (we have very few of them though):

```rust
#[inline] // should be here in case compiler can optimize it out
pub fn foo(val: Bar) {
    assert!(validity_precondition_hard_to_track_in_type_system(val));
    // ...
}
```

If large codegen is a problem, we can just pass the value to `foo_unchecked` function after checking.

**Comment by TheBlueMatt** [comment]
AFAIU `#[inline]` isn't all that interesting internal to a crate (its just a hint, anyway, LLVM may well ignore it), but by default rustc `#[inline]` on `pub` methods means that downstream crates can inline the method, whereas without it the LLVM IR isn't available to do so at all. Thus, `#[inline]`ing any "small" `pub` methods that downstream crates may use in a tight loop seems like a must.

**Reply by Kixunil** [reply]
There's an option to allow inlining of everything (I don't remember what it is) but that slows down compilation a lot, so limiting it to "small" methods is probably better.

**Reply by TheBlueMatt** [reply]
Right, downstream crates can always turn on full LTO, but IIRC that's not true by default in --release, which is what most folks do.
